pkg/capabilities/consensus/ocr3/reporting_plugin.go

```diff
@@ -0,0 +1,232 @@
+package consensus
+
+import (
+	"context"
+
+	ocrcommon "github.com/smartcontractkit/libocr/commontypes"
+	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	"google.golang.org/protobuf/proto"
+
+	pbtypes "github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/ocr3/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+var _ ocr3types.ReportingPlugin[any] = (*reportingPlugin)(nil)
+
+type responder interface {
+	response(ctx context.Context, resp response) error
+}
+
+type reportingPlugin struct {
+	batchSize int
+	s         *store
+	r         responder
+
+	// TODO: make this dynamic based on the request
+	aggregator pbtypes.Aggregator
+	encoder    pbtypes.Encoder
+}
+
+func newReportingPlugin(s *store, r responder, batchSize int) *reportingPlugin {
+	return &reportingPlugin{
+		s:         s,
+		r:         r,
+		batchSize: batchSize,
+	}
+}
+
+func (r *reportingPlugin) Query(ctx context.Context, outctx ocr3types.OutcomeContext) (types.Query, error) {
+	batch, err := r.s.firstN(ctx, r.batchSize)
+	if err != nil {
+		return nil, err
+	}
+
+	ids := []*pbtypes.RequestId{}
+	for _, r := range batch {
+		ids = append(ids, &pbtypes.RequestId{
+			RequestId:   r.RequestID,
+			NamespaceId: r.NamespaceID,
+		})
+	}
+
+	return proto.Marshal(&pbtypes.Query{
+		RequestIds: ids,
+	})
+}
+
+func (r *reportingPlugin) Observation(ctx context.Context, outctx ocr3types.OutcomeContext, query types.Query) (types.Observation, error) {
+	queryReq := &pbtypes.Query{}
+	err := proto.Unmarshal(query, queryReq)
+	if err != nil {
+		return nil, err
+	}
+
+	rids := []string{}
+	for _, q := range queryReq.RequestIds {
+		rids = append(rids, q.RequestId)
+	}
+
+	reqs, err := r.s.getN(ctx, rids)
+	if err != nil {
+		return nil, err
+	}
+
+	obs := &pbtypes.Observation{}
+	for _, r := range reqs {
+		obsPb, err := r.Observations.Proto()
```

Comment:
Side note: I feel that there's no reason for Proto() to ever return an error. If we have a valid Value object, it seems natural that it's always translatable into a valid proto. It looks like right now only Decimal can return an error because we depend on JSON marshaling. Can we fix it somehow and change this method to not return an error at all?

Reply:
Yes, there's probably something we can do to fix this -- I've created a ticket [here](https://smartcontract-it.atlassian.net/jira/software/c/projects/KS/boards/182?selectedIssue=KS-50), and will follow up when I get back from vacation.